File: src/cpp/sim/shared_memory_client.cc
#include "robot_shared_memory.hpp"

#include <chrono>
#include <cstdlib>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

namespace {

struct Args {
  std::string name;
  std::string config = robot_sim::kDefaultConfigPath;
  bool has_torque = false;
  std::vector<double> torque;
  bool disable_command = false;
  double duration = 0.0;
  double rate = 50.0;
};

void PrintUsage(const char* program) {
  std::cerr
      << "Usage: " << program << " [--name NAME] [--config CONFIG]\n"
      << "       [--torque VALUES...] [--disable-command]\n"
      << "       [--duration SECONDS] [--rate HZ]\n";
}

bool IsOption(const std::string& value) {
  return value.rfind("--", 0) == 0;
}

Args ParseArgs(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string option = argv[i];
    auto require_value = [&](const std::string& name) -> std::string {
      if (i + 1 >= argc) {
        throw std::runtime_error(name + " requires a value");
      }
      return argv[++i];
    };

    if (option == "--name") {
      args.name = require_value(option);
    } else if (option == "--config") {
      args.config = require_value(option);
    } else if (option == "--torque") {
      args.has_torque = true;
      while (i + 1 < argc && !IsOption(argv[i + 1])) {
        args.torque.push_back(std::stod(argv[++i]));
      }
    } else if (option == "--disable-command") {
      args.disable_command = true;
    } else if (option == "--duration") {
      args.duration = std::stod(require_value(option));
    } else if (option == "--rate") {
      args.rate = std::stod(require_value(option));
    } else if (option == "-h" || option == "--help") {
      PrintUsage(argv[0]);
      std::exit(0);
    } else {
      throw std::runtime_error("unknown argument: " + option);
    }
  }
  if (args.rate <= 0.0) {
    throw std::runtime_error("--rate must be greater than 0");
  }
  return args;
}

std::string FormatValues(const std::vector<double>& values, std::size_t limit = 6) {
  std::ostringstream output;
  output << "[";
  for (std::size_t i = 0; i < values.size() && i < limit; ++i) {
    if (i != 0) {
      output << ", ";
    }
    output << values[i];
  }
  if (values.size() > limit) {
    output << ", ...";
  }
  output << "]";
  return output.str();
}

std::string RunCycle(robot_sim::RobotSharedMemory& shared_io, const Args& args) {
  robot_sim::RobotState state =
      shared_io.ReadState(std::chrono::duration<double>(1.0));

  if (args.disable_command) {
    shared_io.DisableCommand();
  } else if (args.has_torque) {
    shared_io.WriteTorque(args.torque);
  }

  std::ostringstream output;
  output << "t=" << state.sim_time
         << " alive=" << (state.sim_alive ? 1 : 0);
  for (const auto& field : state.fields) {
    output << " " << field.first << "=" << FormatValues(field.second);
  }
  return output.str();
}

}  // namespace

int main(int argc, char** argv) {
  try {
    const Args args = ParseArgs(argc, argv);
    robot_sim::RobotSharedMemory shared_io =
        robot_sim::RobotSharedMemory::Attach(args.config, args.name);

    if (args.has_torque) {
      const auto it = shared_io.layout().command_index.find("torque");
      if (it == shared_io.layout().command_index.end()) {
        throw std::runtime_error("the JSON config does not define torque");
      }
      const robot_sim::FieldLayout& torque_field =
          shared_io.layout().command_fields[it->second];
      if (args.torque.size() != torque_field.size) {
        throw std::runtime_error(
            "--torque needs " + std::to_string(torque_field.size) +
            " values for this config, got " +
            std::to_string(args.torque.size()));
      }
    }

    const auto deadline =
        std::chrono::steady_clock::now() + std::chrono::duration<double>(args.duration);
    while (true) {
      std::cout << RunCycle(shared_io, args) << "\n";
      if (args.duration <= 0.0 || std::chrono::steady_clock::now() >= deadline) {
        break;
      }
      std::this_thread::sleep_for(std::chrono::duration<double>(1.0 / args.rate));
    }
    return 0;
  } catch (const std::exception& exc) {
    std::cerr << "error: " << exc.what() << "\n";
    PrintUsage(argv[0]);
    return 1;
  }
}

File: src/python/sim/shared_memory_client.py
#!/usr/bin/env python3
"""Example external client for shared-memory robot I/O."""

from __future__ import annotations

import argparse
import time
from pathlib import Path

from robot_shared_memory import DEFAULT_CONFIG_PATH, RobotSharedMemory


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Read robot state and optionally write torque commands."
    )
    parser.add_argument(
        "--name",
        help="Override the shared-memory block name from the JSON config.",
    )
    parser.add_argument(
        "--config",
        type=Path,
        default=DEFAULT_CONFIG_PATH,
        help="JSON config describing shared-memory state and command fields.",
    )
    parser.add_argument(
        "--torque",
        type=float,
        nargs="*",
        help="Torque command in actuator order. Omit to read state only.",
    )
    parser.add_argument(
        "--disable-command",
        action="store_true",
        help="Disable external torque command and send zero torque.",
    )
    parser.add_argument(
        "--duration",
        type=float,
        default=0.0,
        help="Seconds to run. Use 0 for one read/write cycle.",
    )
    parser.add_argument(
        "--rate",
        type=float,
        default=50.0,
        help="Client loop rate in Hz when --duration is greater than 0.",
    )
    parser.add_argument(
        "--log",
        type=Path,
        help="Optional CSV file for sampled state summaries.",
    )
    return parser.parse_args()


def format_values(values: list[float], limit: int = 6) -> str:
    shown = ", ".join(f"{value:.4g}" for value in values[:limit])
    if len(values) > limit:
        shown += ", ..."
    return f"[{shown}]"


def run_cycle(shared_io: RobotSharedMemory, args: argparse.Namespace) -> str:
    state = shared_io.read_state()

    if args.disable_command:
        shared_io.disable_command()
    elif args.torque is not None:
        shared_io.write_torque(args.torque)

    field_summary = " ".join(
        f"{name}={format_values(values)}" for name, values in state.fields.items()
    )
    return (
        f"t={state.sim_time:.4f} "
        f"alive={int(state.sim_alive)} "
        f"{field_summary}"
    )


def main() -> int:
    args = parse_args()
    if args.rate <= 0:
        raise ValueError("--rate must be greater than 0.")

    with RobotSharedMemory.attach(args.name, config_path=args.config) as shared_io:
        if args.torque is not None and "torque" not in shared_io.layout.command_fields:
            raise ValueError("The JSON config does not define a 'torque' command field.")
        if args.torque is not None:
            expected_torque_size = shared_io.layout.command_fields["torque"].size
        if args.torque is not None and len(args.torque) != expected_torque_size:
            raise ValueError(
                f"--torque needs {expected_torque_size} values for this config, "
                f"got {len(args.torque)}."
            )

        log_file = None
        if args.log:
            log_file = args.log.open("w", encoding="utf-8")
            field_names = list(shared_io.layout.state_fields)
            log_file.write(
                ",".join(["wall_time", "sim_time", "sim_alive", *field_names])
                + "\n"
            )

        try:
            deadline = time.monotonic() + args.duration
            while True:
                summary = run_cycle(shared_io, args)
                print(summary)

                if log_file:
                    state = shared_io.read_state()
                    field_values = [
                        f'"{state.fields[name]}"'
                        for name in shared_io.layout.state_fields
                    ]
                    log_file.write(
                        f"{state.wall_time:.9f},{state.sim_time:.9f},"
                        f"{int(state.sim_alive)},"
                        f"{','.join(field_values)}\n"
                    )
                    log_file.flush()

                if args.duration <= 0 or time.monotonic() >= deadline:
                    break
                time.sleep(1.0 / args.rate)
        finally:
            if log_file:
                log_file.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
